### PR TITLE
support fetching origin sync data from a beacon node

### PR DESCRIPTION
**Pull request by kasey**
**What type of PR is this?**
Feature

**What does this PR do? Why is it needed?**

This is PR 2/3 towards implementing checkpoint sync (aka weak subjectivity sync). This PR adds a new RPC endpoint that enable a client to query for the epoch of the latest finalized block within the weak subjectivity period. Also included is a new cli tool `prysmctl` with subcommands enabling a user to download the block at the first slot of the epoch and the subsequent state that integrates the block into the (ie latest_block_header root = the signed beacon block root).

**Other notes for review**

Currently this branch includes some code that modifies the timeout of calls to the state endpoint. This will be removed once https://github.com/prysmaticlabs/prysm/pull/9946 has been merged to develop and updated in this branch.

Note that this branch builds on `cp-sync-origin-root`, so that branch will be used as the target of this PR until the upstream branch is merged, at which time this PR will swap to `develop`.

### CODE REVIEW

File: api/client/openapi/client.go [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+	_, err = b.ReadFrom(r.Body)
+	if err != nil {
+		return nil, err
+	}
+	err = v.UnmarshalSSZ(b.Bytes())
+	return v, err
+}
+
+// GetBlockByRoot retrieves a SignedBeaconBlockAltair with the given root via the beacon node API
+func (c *Client) GetBlockByRoot(blockHex string) (*ethpb.SignedBeaconBlockAltair, error) {
```
Comment (kasey) [inline api/client/openapi/client.go:186]: not using this at the moment, should probably remove it.

File: api/client/openapi/client.go [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+	_, err = b.ReadFrom(r.Body)
+	if err != nil {
+		return nil, err
+	}
+	err = v.UnmarshalSSZ(b.Bytes())
+	return v, err
+}
+
+// GetStateByRoot retrieves a BeaconStateAltair with the given root via the beacon node API
+func (c *Client) GetStateByRoot(stateHex string) (*ethpb.BeaconStateAltair, error) {
```
Comment (kasey) [inline api/client/openapi/client.go:214]: not using this at the moment, should probably remove it.

File: api/client/openapi/client.go [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+	StateRoot string
+	Epoch     string
+}
+
+// GetWeakSubjectivityCheckpoint calls an entirely different rpc method than GetWeakSubjectivityCheckpointEpoch
+// This endpoint is much slower, because it uses stategen to generate the BeaconState at the beginning of the
+// weak subjectivity epoch. This also results in a different set of state+block roots, because this endpoint currently
+// uses the state's latest_block_header for the block hash, while the checkpoint sync code assumes that the block
+// is from the first slot in the epoch and the state is from the subesequent slot.
+func (c *Client) GetWeakSubjectivityCheckpoint() (*ethpb.WeakSubjectivityCheckpoint, error) {
```
Comment (kasey) [inline api/client/openapi/client.go:121]: not using this at the moment, should probably remove it.

File: cmd/prysmctl/checkpoint/save.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+	err = os.WriteFile(blockPath, bb, 0600)
+	if err != nil {
+		return err
+	}
+
+	sb, err := state.MarshalSSZ()
+	if err != nil {
+		return err
+	}
+	//statePath := fmt.Sprintf("state-%s.ssz", fmt.Sprintf("%#x", stateRoot))
```
Comment (rkapka) [inline cmd/prysmctl/checkpoint/save.go:170]: :scream:

File: api/client/openapi/client.go [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+		return nil, err
+	}
+	return &ethpb.WeakSubjectivityCheckpoint{
+		Epoch:     types.Epoch(epoch),
+		BlockRoot: blockRoot,
+		StateRoot: stateRoot,
+	}, nil
+}
+
+// TODO: fix hardcoded pb type using sniff code
```
Comment (rkapka) [inline api/client/openapi/client.go:159]: :scream:

File: api/gateway/apimiddleware/process_request.go [hunk 6]
```diff
@@ -74,10 +74,22 @@ func (m *ApiProxyMiddleware) PrepareRequestForProxying(endpoint Endpoint, req *h
...
+	return func(c *http.Client) {
+		c.Timeout = t
+	}
+}
+
 // ProxyRequest proxies the request to grpc-gateway.
-func ProxyRequest(req *http.Request) (*http.Response, ErrorJson) {
+func ProxyRequest(req *http.Request, opts ...ClientOption) (*http.Response, ErrorJson) {
 	// We do not use http.DefaultClient because it does not have any timeout.
+	// TODO: think about exposing this as a flag, or based on endpoint
```
Comment (rkapka) [inline api/gateway/apimiddleware/process_request.go:88]: :scream:

File: cmd/prysmctl/checkpoint/save.go (hunk 4 above)
Comment (kasey) [reply cmd/prysmctl/checkpoint/save.go:170]: lol, the shaaame

File: api/client/openapi/client.go (hunk 1 above)
Comment (kasey) [reply api/client/openapi/client.go]: removed

File: api/client/openapi/client.go (hunk 2 above)
Comment (kasey) [reply api/client/openapi/client.go]: removed

**Comment by kasey** [comment]
superseded by https://github.com/prysmaticlabs/prysm/pull/10288

**Pull request closed without merging**